src/garage/torch/algos/mtsac.py

```diff
@@ -0,0 +1,208 @@
+            eval_env=eval_envs)
+        self._num_tasks = num_tasks
+        self._eval_envs = eval_envs
+        if self.use_automatic_entropy_tuning and not alpha:
+            if target_entropy:
+                self.target_entropy = target_entropy
+            else:
+                self.target_entropy = -np.prod(
+                    self.env_spec.action_space.shape).item()
+            self.log_alpha = torch.Tensor([self._initial_log_entropy] *
+                                          self._num_tasks).requires_grad_()
+            self.alpha_optimizer = optimizer([self.log_alpha] *
+                                             self._num_tasks,
+                                             lr=self.policy_lr)
+        else:
+            self.log_alpha = torch.Tensor([alpha]).log()
+        self._epoch_mean_success_rate = []
+        self._epoch_median_success_rate = []
+
+    def _get_log_alpha(self, samples_data):
+        """Return the value of log_alpha.
+
+        Args:
+            samples_data (dict): Transitions(S,A,R,S') that are sampled from
+                the replay buffer. It should have the keys 'observation',
+                'action', 'reward', 'terminal', and 'next_observations'.
+
+        Note:
+            samples_data's entries should be numpy.ndarrays with the following
+            shapes:
+                observation:
+                    (sac.env_spec.observation_space, sac.buffer_batch_size)
+                action: (sac.env_spec.action_space, sac.buffer_batch_size)
+                reward: (1, sac.buffer_batch_size)
+                terminal: (1, sac.buffer_batch_size)
+                next_observation:
+                    (sac.env_spec.observation_space, sac.buffer_batch_size)
+
+        Returns:
+            torch.Tensor: log_alpha. shape is (1, self.buffer_batch_size)
+
+        """
+        obs = samples_data['observation']
+        log_alpha = self.log_alpha
+        one_hots = obs[:, -self._num_tasks:]
+        ret = torch.mm(one_hots, log_alpha.unsqueeze(0).t()).squeeze()
+        return ret
+
+    def _evaluate_policy(self, epoch):
+        """Evaluate the performance of the policy via deterministic rollouts.
+
+            Statistics such as (average) discounted return and success rate are
+            recorded.
+
+        Args:
+            epoch(int): The current training epoch.
+
+        Returns:
+            float: The average return across self._num_evaluation_trajectories
+                trajectories
+
+        """
+        eval_trajs = []
+        for env in self._eval_env:
+            eval_trajs.append(
+                self._obtain_evaluation_samples(
+                    env, num_trajs=self._num_evaluation_trajectories))
+        eval_trajs = TrajectoryBatch.concatenate(*eval_trajs)
+        last_return = log_multitask_performance(epoch, eval_trajs,
+                                                self.discount)
+        return last_return
+
+    def to(self, device=None):
```

Comment:
@krzentner is there a way we can easily make this available to all torch algorithms?

Reply:
This mostly "just works" if you just use parameters, but still requires some logic to re-target the optimizer. It's about all we can do for now without re-designing the optimizer API.